packages/@ember/-internals/metal/lib/tracked.ts

```diff
@@ -8,6 +8,168 @@ import { markObjectAsDirty, tagForProperty } from './tags';
+    }
+  };
+
+  warnInAutotrackingTransaction = (fn: () => void) => {
+    if (WARN_IN_AUTOTRACKING_TRANSACTION) {
+      // already in a transaction, continue and let the original transaction finish
+      fn();
+      return;
+    }
+
+    WARN_IN_AUTOTRACKING_TRANSACTION = true;
+
+    try {
+      fn();
+    } finally {
+      WARN_IN_AUTOTRACKING_TRANSACTION = false;
+    }
+  };
+
+  let nthIndex = (str: string, pattern: string, n: number, startingPos = -1) => {
+    let i = startingPos;
+
+    while (n-- > 0 && i++ < str.length) {
+      i = str.indexOf(pattern, i);
+      if (i < 0) break;
+    }
+
+    return i;
+  };
+
+  let makeAutotrackingErrorMessage = (
+    sourceData: AutotrackingTransactionSourceData,
+    obj: object,
+    keyName?: string
+  ) => {
+    let dirtyString = keyName
+      ? `\`${keyName}\` on \`${getDebugName!(obj)}\``
+      : `\`${getDebugName!(obj)}\``;
+
+    let message = [
+      `You attempted to update ${dirtyString}, but it had already been used previously in the same computation.  Attempting to update a value after using it in a computation can cause logical errors, infinite revalidation bugs, and performance issues, and is not supported.`,
+    ];
+
+    if (sourceData.context) {
+      message.push(`\`${keyName}\` was first used when rendering:`);
+      message.push(sourceData.context.replace(/^/gm, '    '));
+    }
+
+    if (sourceData.error.stack) {
+      let sourceStack = sourceData.error.stack;
+      let thirdIndex = nthIndex(sourceStack, '\n', 3);
+      sourceStack = sourceStack.substr(thirdIndex);
+
+      message.push(`Stack trace for the first usage: ${sourceStack}`);
+    }
+
+    message.push(`Stack trace for the update:`);
+
+    return message.join('\n\n');
+  };
+
+  setRenderingContextDesc = (desc: string) => {
+    renderingContextDescription = desc;
+  };
+
+  clearRenderingContextDesc = () => {
+    renderingContextDescription = '';
+  };
+
+  markTagAsConsumed = (_tag: Tag, sourceError: Error) => {
+    if (!AUTOTRACKING_TRANSACTION || AUTOTRACKING_TRANSACTION.has(_tag)) return;
+
+    AUTOTRACKING_TRANSACTION.set(_tag, {
+      context: renderingContextDescription,
+      error: sourceError,
+    });
+
+    // We need to mark the tag and all of its subtags as consumed, so we need to
+    // cast in and access its internals. In the future this shouldn't be necessary,
+    // this is only for computed properties.e
+    let tag = _tag as any;
+
+    if (tag.subtag) {
+      markTagAsConsumed!(tag.subtag, sourceError);
+    }
+
+    if (tag.subtags) {
+      tag.subtags.forEach((tag: Tag) => markTagAsConsumed!(tag, sourceError));
+    }
+  };
+
+  assertTagNotConsumed = (tag: Tag, obj: object, keyName?: string, forceHardError = false) => {
+    if (AUTOTRACKING_TRANSACTION === null) return;
+
+    let sourceData = AUTOTRACKING_TRANSACTION.get(tag);
+
+    if (!sourceData) return;
+
+    if (WARN_IN_AUTOTRACKING_TRANSACTION && !forceHardError) {
```

Comment:
Hmm, so my intention was that `warnInAutotrackingTransaction` should only ever be used within an existing transaction, it would not start one. I can add an assertion to the helper to tell us if we're using it wrong.

Reply:
>  I can add an assertion to the helper to tell us if we're using it wrong.

Ya, that sounds great. Thank you for clarifying!